src/util/requester.ts

```diff
@@ -0,0 +1,281 @@
+    if (timeToWait) {
+      logger.trace(
+        `Ran into request limits, sleeping ${timeToWait}ms before executing (Key: ${next.key}, Retry #: ${next.retries})`,
+      )
+      await sleep(timeToWait)
+    }
+
+    // Fire off to complete in the background. We don't wait here to be able to fire off multiple requests concurrently
+    this.executeRequest.bind(this)(next)
+
+    return this.processNext()
+  }
+
+  // Handler for the requests that will be fired off, eventually resolving the promise associated with the queued request
+  private async executeRequest(req: QueuedRequest) {
+    const { key, config, resolve, reject, retries } = req
+    const providerDataRequested = Date.now()
+    const responseTimer = metrics.dataProviderRequestDurationSeconds.startTimer()
+
+    // Set configured timeout for all requests unless manually specified
+    config.timeout = config.timeout || this.timeout
+
+    try {
+      logger.trace(`Sending request (Key: ${key}) to data provider`)
+      const response = await axios.request(config)
+      logger.trace(`Request (Key: ${key}) was successful `)
+      resolve({
+        response,
+        timestamps: {
+          providerDataRequested,
+          providerDataReceived: Date.now(),
+        },
+      })
+
+      // Record count of successful data provider requests
+      metrics.dataProviderRequests
+        .labels(metrics.dataProviderMetricsLabel(response.status, config.method))
+        .inc()
+    } catch (e) {
+      if (retries >= this.maxRetries) {
+        logger.trace(`Request failed and no more retries remaining, rejecting promise...`)
+        const err = e as AxiosError
+        const ErrorClass = err.response?.status ? AdapterDataProviderError : AdapterConnectionError
+
+        // Record count of failed data provider request
+        metrics.dataProviderRequests
+          .labels(metrics.dataProviderMetricsLabel(err.response?.status || 0, config.method))
+          .inc()
+
+        reject(
+          new ErrorClass(
+            {
+              statusCode: 502,
+              name: 'Data Provider error',
+              providerStatusCode: err?.response?.status ?? 502,
+              message: err?.message,
+              cause: e,
+              errorResponse: err?.response?.data,
+              url: config.url,
+            },
+            {
+              providerDataRequested,
+              providerDataReceived: Date.now(),
+              providerIndicatedTime: undefined,
+            },
+          ),
+        )
+      } else {
+        const timeToSleep = (2 ** retries + Math.random()) * 1000
+        logger.trace(`Request failed, sleeping for ${timeToSleep}ms...`)
+        await sleep(timeToSleep)
+
+        req.retries++
+        logger.trace(`Adding request to the queue (Key: ${key}, Retry #: ${req.retries})`)
+        this.queue.add(req)
+      }
+    } finally {
+      // Remove the request from our map
+      delete this.map[key]
```

Comment:
If I'm understanding the logic correctly, wouldn't we want to only remove from the map if we've hit retries limit? We'd still want the request to be in the map if we'd added it to the back of the queue for a retry right?

Reply:
correct, I'll make the change